buffer/src/test/java/io/netty/buffer/AbstractByteBufTest.java

```diff
@@ -2910,6 +2919,78 @@ public void run() {
         assertNull(cause.get());
     }
 
+    public static Stream<Arguments> setCharSequenceCombinations() {
+        Stream.Builder<Arguments> builder = Stream.builder();
+        List<Charset> charsets = Arrays.asList(
+                StandardCharsets.UTF_8,
+                StandardCharsets.US_ASCII,
+                StandardCharsets.ISO_8859_1);
+        for (Charset charset : charsets) {
+            for (CharSequenceType charSequenceType : CharSequenceType.values()) {
+                builder.add(Arguments.of(charset, charSequenceType));
+            }
+        }
+        return builder.build();
+    }
+
+    enum CharSequenceType {
+        STRING,
+        ASCII_STRING;
+
+        public CharSequence create(char[] cs) {
+            switch (this) {
+                case STRING:
+                    return new String(cs);
+                case ASCII_STRING:
+                    return new AsciiString(cs);
+            }
+            throw new UnsupportedOperationException("Unknown type: " + this);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    @ParameterizedTest
+    @MethodSource("setCharSequenceCombinations")
+    void testSetCharSequenceMultipleThreads(final Charset charset, CharSequenceType charSeqType) throws Exception {
+        int bufSize = 32;
+        ByteBuf[] bufs = new ByteBuf[16];
+        for (int i = 0; i < bufs.length; i++) {
+            bufs[i] = newBuffer(bufSize);
+        }
+
+        int iterations = 256;
+        Semaphore start = new Semaphore(0);
+        Semaphore finish = new Semaphore(0);
+        char[] cs = new char[(int) (bufSize / charset.newEncoder().maxBytesPerChar())];
+        Arrays.fill(cs, 'a');
+        final CharSequence str = charSeqType.create(cs);
+        ExecutorService executor = Executors.newFixedThreadPool(bufs.length);
+        try {
+            Future<Void>[] futures = new Future[bufs.length];
+            for (int i = 0; i < bufs.length; i++) {
+                final ByteBuf buf = bufs[i];
+                futures[i] = executor.submit(() -> {
+                    finish.release();
+                    start.acquire();
+                    for (int j = 0; j < iterations; j++) {
+                        buf.setCharSequence(0, str, charset);
+                    }
+                    return null;
+                });
+            }
+            finish.acquire(bufs.length);
+            start.release(bufs.length);
+            for (Future<Void> future : futures) {
+                future.get();
```

Comment:
Shouldn't we await first on all and then check if any of them is failed ? Otherwise we might have stuff still running while shutdown the executor

Reply:
What I wanted to say is that I think we should catch the ExecutionException and only rethrow if we called get on all futures